### PR TITLE
[SPARK-23435][SPARKR][TESTS] Update testthat to >= 2.0.0 

### PR DESCRIPTION
**Pull request by zero323**
### What changes were proposed in this pull request?

- Update `testthat` to >= 2.0.0
- Replace of `testthat:::run_tests` with `testthat:::test_package_dir`
- Add trivial assertions for tests, without any expectations, to avoid skipping.
- Update related docs.

### Why are the changes needed?

`testthat` version has been frozen by [SPARK-22817](https://issues.apache.org/jira/browse/SPARK-22817) / https://github.com/apache/spark/pull/20003, but 1.0.2 is pretty old, and we shouldn't keep things in this state forever.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

- Existing CI pipeline:
     - Windows build on AppVeyor, R 3.6.2, testthtat 2.3.1
     - Linux build on Jenkins, R 3.1.x, testthat 1.0.2 

- Additional builds with thesthat 2.3.1  using [sparkr-build-sandbox](https://github.com/zero323/sparkr-build-sandbox) on c7ed64af9e697b3619779857dd820832176b3be3 

   R 3.4.4  (image digest ec9032f8cf98)
   ```
   docker pull zero323/sparkr-build-sandbox:3.4.4
   docker run zero323/sparkr-build-sandbox:3.4.4 zero323 --branch SPARK-23435 --commit c7ed64af9e697b3619779857dd820832176b3be3 --public-key https://keybase.io/zero323/pgp_keys.asc
    ```
    3.5.3 (image digest 0b1759ee4d1d)

    ```
    docker pull zero323/sparkr-build-sandbox:3.5.3
    docker run zero323/sparkr-build-sandbox:3.5.3 zero323 --branch SPARK-23435 --commit 
    c7ed64af9e697b3619779857dd820832176b3be3 --public-key https://keybase.io/zero323/pgp_keys.asc
    ```

   and 3.6.2 (image digest 6594c8ceb72f)
    ```
   docker pull zero323/sparkr-build-sandbox:3.6.2
   docker run zero323/sparkr-build-sandbox:3.6.2 zero323 --branch SPARK-23435 --commit c7ed64af9e697b3619779857dd820832176b3be3 --public-key https://keybase.io/zero323/pgp_keys.asc
   ````

   Corresponding [asciicast](https://asciinema.org/) are available as 10.5281/zenodo.3629431

     [DOI](https://doi.org/10.5281/zenodo.3629431) 

   (a bit to large to burden asciinema.org, but can run locally via `asciinema play`).

----------------------------

Continued from #27328

### CODE REVIEW

**Comment by zero323** [comment]
Assuming that Windows test passes (it did previously), it is ready for review.

Unfortunately I couldn't figure out how to correctly install testthat 2.0 on Jenkins from inside test runner ‒ some configurations that worked locally (on official R-base docker image with 3.1.3, still failed here), and I don't want to throttle Jenkins anymore with my lousy attempts (sorry for that).

 So I decided to add fallback case, so it runs for now on both testthat 1.x and 2.x. It means that appveyor runs with 2.x and Jenkins with 1.x ‒ I guess it might be considered a good thing for now (@shaneknapp?). The only caveat is that we'll have to pay more attention to  AppVeyor builds, until Jenkins is upgraded, as it has become clear, that current setup can silence some errors.

CC @HyukjinKwon @felixcheung

File: R/pkg/tests/fulltests/test_sparkSQL.R [hunk 1]
```diff
@@ -848,24 +848,31 @@ test_that("collect() and take() on a DataFrame return the same number of rows an
...
+  lines <- readLines(jsonPath, encoding = "UTF-8")
 
-  jsonPath <- tempfile(pattern = "sparkr-test", fileext = ".tmp")
-  writeLines(lines, jsonPath)
+  expected <- regmatches(lines, gregexpr('(?<="name": ").*?(?=")', lines, perl = TRUE))
 
   df <- read.df(jsonPath, "json")
   rdf <- collect(df)
   expect_true(is.data.frame(rdf))
-  expect_equal(rdf$name[1], markUtf8("안녕하세요"))
```
Comment (dongjoon-hyun) [inline R/pkg/tests/fulltests/test_sparkSQL.R]: Ur, this is a required change for `testthat`?

File: R/pkg/tests/fulltests/test_sparkSQL.R [hunk 2]
```diff
@@ -848,24 +848,31 @@ test_that("collect() and take() on a DataFrame return the same number of rows an
...
 test_that("collect() support Unicode characters", {
-  lines <- c("{\"name\":\"안녕하세요\"}",
-             "{\"name\":\"您好\", \"age\":30}",
-             "{\"name\":\"こんにちは\", \"age\":19}",
-             "{\"name\":\"Xin chào\"}")
+  jsonPath <- file.path(
+    Sys.getenv("SPARK_HOME"),
+    "R", "pkg", "tests", "fulltests", "data",
+    "test_utils_utf.json"
+  )
```
Comment (dongjoon-hyun) [inline R/pkg/tests/fulltests/test_sparkSQL.R]: This new change looks irrelevant to `Update test that to >= 2.0.0`. I'd like to recommend you to revert this change.

File: R/pkg/tests/fulltests/test_utils.R [hunk 3]
```diff
@@ -89,7 +89,10 @@ test_that("cleanClosure on R functions", {
     lapply(x, g) + 1  # Test for capturing function call "g"'s closure as a argument of lapply.
     l$field[1, 1] <- 3  # Test for access operators `$`.
     res <- defUse + l$field[1, ]  # Test for def-use chain of "defUse", and "" symbol.
-    f(res)  # Test for recursive calls.
+    # Enable once SPARK-30629 is fixed
```
Comment (dongjoon-hyun) [inline R/pkg/tests/fulltests/test_utils.R]: Does this fail with `testthat >= 2.0.0`?

Comment (zero323) [reply R/pkg/tests/fulltests/test_utils.R]: @dongjoon-hyun It fails on any version, but is somehow silenced if we use testthat 1.x. I've already opened a JIRA ticket for ‒ [SPARK-30629](https://issues.apache.org/jira/browse/SPARK-30629). Can be easily reproduced outside tests.

File: R/pkg/tests/fulltests/test_sparkSQL.R (hunk 1 above)
Comment (zero323) [reply R/pkg/tests/fulltests/test_sparkSQL.R]: This fails on Windows due to encoding issues, irrespective of testthat version. Similar to recursive call test case, failure is easy to reproduce (Windows, English locale,  CP1252 encoding) outside tests, but somehow is ignored in testthat 1.x.

@dongjoon-hyun The alternative to this is to write bytes explicitly, i.e.

```
test_that("collect() support Unicode characters", {

  lines <- markUtf8(c(
    '{"name": "안녕하세요"}',
    '{"name": "您好", "age": 30}',
    '{"name": "こんにちは", "age": 19}',
    '{"name": "Xin ch\xc3\xa0o"}'
  ))

  jsonPath <- tempfile(pattern = "sparkr-test", fileext = ".tmp")
  writeLines(lines, jsonPath, useBytes = TRUE)

  expected <- regmatches(lines, regexec('(?<="name": ").*?(?=")', lines, perl = TRUE))

  df <- read.df(jsonPath, "json")
  rdf <- collect(df)
  expect_true(is.data.frame(rdf))

  rdf$name <- markUtf8(rdf$name)
  expect_equal(rdf$name[1], expected[[1]])
  expect_equal(rdf$name[2], expected[[2]])
  expect_equal(rdf$name[3], expected[[3]])
  expect_equal(rdf$name[4], expected[[4]])

  df1 <- createDataFrame(rdf)
  expect_equal(
    collect(
      where(df1, df1$name == expected[[2]])
    )$name,
    expected[[2]]
  )
})
```

or skip test case on Windows, but as-is, this will and should fail on AppVeyor.

File: R/pkg/tests/fulltests/test_sparkSQL.R (hunk 2 above)
Comment (zero323) [reply R/pkg/tests/fulltests/test_sparkSQL.R]: Ditto.

File: R/pkg/tests/fulltests/test_sparkSQL.R (hunk 1 above)
Comment (dongjoon-hyun) [reply R/pkg/tests/fulltests/test_sparkSQL.R]: Please make another PR for that. One PR should have one theme.
>  irrespective of testthat version

File: R/pkg/tests/fulltests/test_sparkSQL.R (hunk 2 above)
Comment (dongjoon-hyun) [reply R/pkg/tests/fulltests/test_sparkSQL.R]: Ditto.

File: R/pkg/tests/fulltests/test_utils.R (hunk 3 above)
Comment (dongjoon-hyun) [reply R/pkg/tests/fulltests/test_utils.R]: Let's fix SPARK-30629 first.

File: R/pkg/tests/fulltests/test_sparkSQL.R (hunk 1 above)
Comment (zero323) [reply R/pkg/tests/fulltests/test_sparkSQL.R]: So you're saying that we should accept invalid test failures here? Or fix this beforehand on 1.x, even if problem is not visible with 1.x?

Comment (dongjoon-hyun) [reply R/pkg/tests/fulltests/test_sparkSQL.R]: Yes, right. That would be great.

Comment (dongjoon-hyun) [reply R/pkg/tests/fulltests/test_sparkSQL.R]: For Windows, @HyukjinKwon is the best person. So, he can override my comment.

Comment (zero323) [reply R/pkg/tests/fulltests/test_sparkSQL.R]: Do you have any preference towards the method used?

In general I have no access to Windows setup on which I can test this across testthat (things vary, as 2.0 and latest 2.x versions come with different dependencies), R (we have 3.1 on Jenkins, latest on Windows builds) and encoding. So this is the best I came up with, in the limited time I had to tinker with the whole thing. But I am opened to other proposals.

Comment (dongjoon-hyun) [reply R/pkg/tests/fulltests/test_sparkSQL.R]: Ur, if you are unable to verify, please don't make a PR.
> I have no access to Windows setup

File: R/pkg/tests/fulltests/test_utils.R (hunk 3 above)
Comment (zero323) [reply R/pkg/tests/fulltests/test_utils.R]: OK. Set SPARK-30629  as a blocker for SPARK-23435.

Comment (dongjoon-hyun) [reply R/pkg/tests/fulltests/test_utils.R]: Thanks, @zero323 . SPARK-30629 only happens at test that >= 2.0.0. It should be a blocker for this (SPARK-23435).

File: R/pkg/tests/fulltests/test_sparkSQL.R (hunk 1 above)
Comment (zero323) [reply R/pkg/tests/fulltests/test_sparkSQL.R]: For the reference I've opened [SPARK-30645](https://issues.apache.org/jira/browse/SPARK-30645) to track this one.

File: R/pkg/tests/fulltests/test_utils.R (hunk 3 above)
Comment (zero323) [reply R/pkg/tests/fulltests/test_utils.R]: Dropping this commit from here.

**Comment by HyukjinKwon** [comment]
I am all good as long as we can use the latest `testthat`. This is a really annoying and rather critical problem because many R libraries nowsdays has `testthat` dependency.

**Comment by zero323** [comment]
@HyukjinKwon 

> I am all good as long as we can use the latest `testthat`. This is a really annoying and rather critical problem because many R libraries nowsdays has `testthat` dependency.

Following @dongjoon-hyun advice I moved commits fixing test issues discovered here out this PR. Could you take a look at https://github.com/apache/spark/pull/27362 and see if it makes sense to you?

As of the other one, I'll try to debug this over next few days and see if I can find a fix, if not questions remains if it should be a blocker for testthat upgrade. It feel like a bit like sweeping things under the rug.

**Comment by HyukjinKwon** [comment]
> Unfortunately I couldn't figure out how to correctly install testthat 2.0 on Jenkins from inside test runner ‒ some configurations that worked locally (on official R-base docker image with 3.1.3, still failed here), and I don't want to throttle Jenkins anymore with my lousy attempts (sorry for that).

@zero323, so basically you couldn't figure out CRAN check to work with R 3.1.3? If that's the case, does it work with the latest of R 3.4.x (at least in your local)?

Spark will deprecate R lower versions than 3.4 in Spark 3.0 and we will drop it right away after Spark 3.0 release. So, R version in Jenkins will increase to the latest of 3.4.x ideally.

File: R/pkg/tests/run-all.R [hunk 4]
```diff
@@ -60,11 +59,23 @@ if (identical(Sys.getenv("NOT_CRAN"), "true")) {
   if (identical(Sys.getenv("NOT_CRAN"), "true")) {
     # set random seed for predictable results. mostly for base's sample() in tree and classification
     set.seed(42)
-    # for testthat 1.0.2 later, change reporter from "summary" to default_reporter()
-    testthat:::run_tests("SparkR",
-                         file.path(sparkRDir, "pkg", "tests", "fulltests"),
-                         NULL,
-                         "summary")
+
+    # To be removed once testthat 1.x is removed from all builds
```
Comment (HyukjinKwon) [inline R/pkg/tests/run-all.R]: +1. I like the idea of keeping it to make the dev env change smoother. Also, we could decouple this PR and Jenkins R upgrade issue.

**Comment by zero323** [comment]
> @zero323, so basically you couldn't figure out CRAN check to work with R 3.1.3? If that's the case, does it work with the latest of R 3.4.x (at least in your local or any Linux based OS)?

As long as fix for [SPARK-30645](https://issues.apache.org/jira/browse/SPARK-30645) is applied and test for cleanClosure on recursive ([SPARK-30629](https://issues.apache.org/jira/browse/SPARK-30629)) is disabled everything looks green, i.e.

- Passes all tests locally on R 3.6.2 / testthat 2.3.1
- It passes all tests on AppVeyor (I don't recall R version there, but testthat is 2.3.1).
- On Jenkins we fallback to 1.x so it is not really meaningful (I don't see how SPARK-30629 flies under the radar there), but all tests passed as well.

About _I couldn't figure_ ‒ I tried to `install.packages` inside `R/pkg/tests/run-all.R` in different combinations that seem to work locally, but it failed miserably on Jenkins (in general with R 3.1 used there we can target only `testthat` 2.0 or so, as later versions require 0.4, which in turn depends on R >= 3.2).

File: R/pkg/tests/fulltests/test_utils.R (hunk 3 above)
Comment (zero323) [reply R/pkg/tests/fulltests/test_utils.R]: CC @falaki

**Comment by HyukjinKwon** [comment]
@zero323, do you mind if I ask to check R 3.4.x latest and `testthat` latest combination? After we drop R lower than 3.4, we will probably switch R version to 3.4.x latest in Jenkins. Then, I think this PR is almost good to go after https://issues.apache.org/jira/browse/SPARK-30629. I, @felixcheung or @shivaram might have to take another look though.

**Comment by zero323** [comment]
> @zero323, do you mind if I ask to check R 3.4.x latest and `testthat` latest combination? After we drop R lower than 3.4 (which will be after Spark 3.0 release since we already deprecated), we will probably switch R version to 3.4.x latest in Jenkins. Then, I think this PR is almost good to go after https://issues.apache.org/jira/browse/SPARK-30629. I, @felixcheung or @shivaram might have to take another look though.

Not at all, unless you ask for Windows build :) 

Seriously though, I had enough problems in category "works for me" with this thing, to be careful. Once https://github.com/apache/spark/pull/27363 I'll rebase, and make another local build in docker to be sure.

File: appveyor.yml [hunk 5]
```diff
@@ -43,11 +43,8 @@ install:
   - ps: .\dev\appveyor-install-dependencies.ps1
   # Required package for R unit tests
   - cmd: R -e "install.packages(c('knitr', 'rmarkdown', 'e1071', 'survival', 'arrow'), repos='https://cloud.r-project.org/')"
-  # Here, we use the fixed version of testthat. For more details, please see SPARK-22817.
-  # As of devtools 2.1.0, it requires testthat higher then 2.1.1 as a dependency. SparkR test requires testthat 1.0.2.
-  # Therefore, we don't use devtools but installs it directly from the archive including its dependencies.
   - cmd: R -e "install.packages(c('crayon', 'praise', 'R6'), repos='https://cloud.r-project.org/')"
-  - cmd: R -e "install.packages('https://cloud.r-project.org/src/contrib/Archive/testthat/testthat_1.0.2.tar.gz', repos=NULL, type='source')"
+  - cmd: R -e "install.packages('testthat', repos='https://cloud.r-project.org/')"
```
Comment (dongjoon-hyun) [inline appveyor.yml]: Shall we add `testthat` to line 46 (https://github.com/apache/spark/pull/27359/files#diff-180360612c6b8c4ed830919bbb4dd459R46)?

File: R/pkg/tests/run-all.R [hunk 6]
```diff
@@ -60,11 +59,23 @@ if (identical(Sys.getenv("NOT_CRAN"), "true")) {
...
+      reporter <- "summary"
+
+    } else {
+      # testthat >= 2.0.0
+      test_runner <- testthat:::test_package_dir
+      reporter <- testthat::default_reporter()
+    }
+
+    test_runner("SparkR",
+           file.path(sparkRDir, "pkg", "tests", "fulltests"),
```
Comment (dongjoon-hyun) [inline R/pkg/tests/run-all.R]: Indentation?

File: appveyor.yml (hunk 5 above)
Comment (HyukjinKwon) [reply appveyor.yml]: +1 for @dongjoon-hyun's advice.

File: R/pkg/tests/run-all.R (hunk 6 above)
Comment (HyukjinKwon) [reply R/pkg/tests/run-all.R]: Should be:

```r
    test_runner("SparkR",
                file.path(sparkRDir, "pkg", "tests", "fulltests"),
                ...
```

to be consistent.

**Comment by dongjoon-hyun** [comment]
@zero323 . Please update the PR description. For example, the followings?
```
Disables unit tests for faulty code (SPARK-30629)
Modify Unicode handling test case to be handled properly independent of OS.
```

**Comment by zero323** [comment]
@HyukjinKwon 

> @zero323, do you mind if I ask to check R 3.4.x latest and testthat latest combination

I'd say we're good:

[asciicast](https://asciinema.org/a/xiIOy6OcntE6hxNXQwI7vcUl0)

Additionally to local builds this gives us:

- R 3.1.x, `testthat` 1.0.2 on Linux (Jenkins)
- R 3.4.3, `testthat` 2.3.1 on Linux (docker build)
- R 3.6.2, `testthtat` 2.3.1 on Windows

**Comment by zero323** [comment]
> Please update the PR description. For example, the followings?

All done @dongjoon-hyun

File: R/pkg/tests/run-all.R (hunk 4 above)
Comment (dongjoon-hyun) [reply R/pkg/tests/run-all.R]: +1. BTW, we had better file a new JIRA and make this comment as an IDed TODO like the following.
```
# TODO(SPARK-XXXXX) To be removed once testthat 1.x is removed from all builds
```

**Comment by dongjoon-hyun** [comment]
@zero323 . Thank you for the screencast. However, it skipped all arrow related tests.

**Comment by zero323** [comment]
> @zero323 . Thank you for the screencast. However, it skipped all arrow related tests. Please playback the screencast.

Unfortunately R arrow is not standalone package (like Python one) and it requires system packages with C++ bindings (installing `arrow` package is not sufficient), And that's dependency hell as these R images (there still more stable than R-base ones) are not really designed for updates. So skipped Arrow tests are expected.

I don't think that really affects the results though, as primary concern was CRAN tests and overall process, and Arrow related code hasn't been touched.

**Comment by dongjoon-hyun** [comment]
We cannot say `We're good` when we know something wrong. I'd like to recommend you to mention what you've done clearly. That's enough. Please note that I'm supporting your effort on this PR. Otherwise, I'll not chim in here to add comments.
> So skipped Arrow tests are expected.

**Comment by zero323** [comment]
> Please note that I'm supporting your effort on this PR. Otherwise, I'll not chim in here to add comments.

Thank you, I appreciate that. 

In general, full reproducible is defined by the Dockerfile which is shown at the begging,  but to put it here for reference 

```
FROM rocker/verse:3.4.3
RUN apt-get update \
    && apt-get install -y --no-install-recommends gpg openjdk-8-jdk-headless \
    && apt-get clean \
    && rm -rf /var/lib/apt/lists/*ce

RUN wget -qO- https://keybase.io/zero323/pgp_keys.asc | gpg --import
RUN git clone --depth 1 --branch SPARK-23435 https://github.com/zero323/spark.git
WORKDIR spark
RUN git rev-parse HEAD
RUN git verify-commit -v HEAD
RUN build/mvn -DskipTests -Phive -Psparkr clean package
RUN R --version
RUN R -e "install.packages(c('e1071', 'praise'))"
RUN R -e "install.packages('testthat', repos='https://cloud.r-project.org/'); packageVersion('testthat'); sessionInfo()"
RUN R/create-rd.sh
RUN R/create-docs.sh
RUN R/check-cran.sh
RUN R/run-tests.sh
```
It can be re-run to confirm that it reflects current state of things.

As show in the cast, build are done directly from this head of this branch (signature is verified) and no changes to the codebase, beyond what is proposed in this PR (and we don't touch any Arrow related components here at all).

As of skipping Arrow tests - that's default behavior defined in respective test for example here

https://github.com/apache/spark/blob/43d9c7e7e57749ee611e0c97781a71a0645b5e9b/R/pkg/tests/fulltests/test_sparkSQL_arrow.R#L25

and following lines. So it is neither failure or result of any source modification. 

Can we make arrow tests run? Possibly, but:

- R Arrow package is not present in snapshot repositories used by rocker images. Installing testthat from https://cloud.r-project.org, already pushed things a lot. Additionally some transitive dependencies have hidden version bounds.
- C++ Arrow bindings would require external system repositories, which can break decencies for R.
- Using other images (let's say official R-base) is not an option, as we need Tex as well as OpenSSL and Curl dev libraries and this will either break or require update of R beyond 2.4 (at least it did for other build configurations I considered).

At this point Spark has no coverage for any intermediate R version (Jenkins runs 3.1 and then we have almost eight years of releases worth gap to 3.6 on AppVeyor), not to mention version-OS combinations. That's troubling, and as work related to this PR shown, can miss obvious errors.  but  not something that can be really addressed by running ad-hoc tests outside project infrastructure.

Anyway... If you have specific concerns about the process used here, and you suspect that proposed changes can lead to problems in the future, I'll do my best to address these.

**Comment by dongjoon-hyun** [comment]
? @zero323 . It seems that you missed my point. I advised like the following.
>  I'd like to recommend you to mention what you've done clearly. That's enough.

Let me rephrase my words. "In the PR description, write that you didn't run the full test. Especially Arrow tests are skipped". It was my request.

**Review by HyukjinKwon — APPROVED** [review]
LGTM.

Arrow is a valid point although AppVeyor (with R latest and CRAN disabled) tests it. It should be best just to test it out if that's feasible considering the importance of this PR and that changing env in Jenkins is difficult to revert if there's anything wrong.

If there's some technical difficulties, I am okay without testing out because the required codes change `expect_true(TRUE)` in the tests seem not required for Arrow tests logically at least.

**Review by HyukjinKwon — COMMENTED** [review]
Okay, given https://github.com/apache/spark/pull/27359#issuecomment-578581851, I am okay with not trying Arrow installation out. 

@shaneknapp made it installing Arrow on R 3.1 (surprisingly ..). Although it's a bit of a while ago and Arrow R was at the early stage, I remember I badly failed to install Arrow in some low R versions.

I think we can verify it later when we actually upgrade R version. Seems it should be easy to fix if there'a anything wrong with Arrow tests.

Also, I found a test case that has a similar structure with Arrow test case as a reference:
https://github.com/apache/spark/blob/d69ed9afdf2bd8d03aaf835292b92692ec8189e9/R/pkg/tests/fulltests/test_sparkSQL.R#L3327-L3376

So I suspect it's good enough to go (except one minor comment https://github.com/apache/spark/pull/27359#discussion_r371046762)

File: R/pkg/tests/run-all.R (hunk 4 above)
Comment (HyukjinKwon) [reply R/pkg/tests/run-all.R]: @zero323, can you handle this comment? Otherwise, It think it's good to go.

**Comment by zero323** [comment]
@dongjoon-hyun 

> ? @zero323 . It seems that you missed my point. I advised like the following.
> 
> > I'd like to recommend you to mention what you've done clearly. That's enough.
> 
> Let me rephrase my words. "In the PR description, write that you didn't run the full test. Especially Arrow tests are skipped". It was my request.

Seems like. I guess there is something lost in translation here

> We cannot say We're good when we know something wrong

Especially what we consider "wrong" in this context, hence my elaborate response. Which I think is still valid ‒ overall test matrix is tiny compared to what is officially supported. Sadly Arrow makes things worse as

- It is not Maven dependency.
- Building from source is pain, and repositories don't provide archival versions.
- Depending on OS, there are also additional dependencies.
- Breaking API changes are not unusual (for example it changed between first 3.0.0 preview and the second), leading to rather [cryptic errors](https://twitter.com/_zero323/status/1202005430471450625).

It would be great if we could either isolate that (I played a bit with Anaconda for that, and results are quite promising, but I am not there yet, not to mention it won't help with all the tech. debt caused by supporting long outdated R and Python versions), or find a way to bundle within Maven builds.

For now I published [zero323/sparkr-build-sandbox](https://github.com/zero323/sparkr-build-sandbox) ([docker.com/r/zero323/sparkr-build-sandbox](https://hub.docker.com/r/zero323/sparkr-build-sandbox)) - not exactly what is needed, but does the trick for now.

Please check updated description and let me know if that addresses your concerns.

File: R/pkg/tests/run-all.R (hunk 4 above)
Comment (zero323) [reply R/pkg/tests/run-all.R]: @HyukjinKwon Done 56b1ba966467f5677572d70fa5c4d75bc0c66789

**Comment by dongjoon-hyun** [comment]
Sorry, but it doesn't address my concerns.

> Please check updated description and let me know if that addresses your concerns.

Screen Shot 2020-01-28 at 2 14 29 PM

I ask you the following. In the above, I cannot see the following content. Even, you completely hide the word `arrow`.
> "In the PR description, write that you didn't run the full test. Especially Arrow tests are skipped". It was my request.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
@HyukjinKwon . Sorry, but you know that I'm not against that @zero323 skipped some tests due to the difficulty. I fully understand that it's sometime unfair to ask that to the contributors. However, the author of PR should mention what he've done in the PR description correctly.

I'm not sure why @zero323 adds many comments about this difficulty, but he didn't mention any of those difficulty into the PR description which will become a permanent commit log. Could you advise him about that? I already add the same comments repeatedly. So, it's too much for me.

Or, @HyukjinKwon . You can merge this PR as is. It's up to you. I'll not be against your decision.

**Comment by zero323** [comment]
@dongjoon-hyun 

> "In the PR description, write that you didn't run the full test. 

At this point all tests *did* run across multiple additional R versions. This is both documented in a persistent and verifiable way, and reproducible.

**Comment by zero323** [comment]
Retest this please.

**Comment by HyukjinKwon** [comment]
@zero323, mind adding a couple of words to describe `arrow` was not installed due to technical difficulties, into the PR description? Seems like there was some miscommunication here.

It's fine to describe what happened during manual test. No one can't be flawless during such tests; it's better to describe it thoroughly for people who follow up later.

**Comment by zero323** [comment]
@HyukjinKwon 

> mind adding a couple of words to describe arrow was not installed due to technical difficulties, into the PR description?

I was actually able to put all the dependencies together, including Arrow. 

> Seems like there was some miscommunication here.

Indeed. I guess I should have skipped explanations.

Anyway, I am happy to close this, and leave for someone more articulate to continue. Everything should be in good shape here from technical perspective, and linked images should provide good starting point for anyone willing to continue. Let's just see if AppVeyor failure is a false positive. Seems like [the last run](https://ci.appveyor.com/project/ApacheSoftwareFoundation/spark/builds/30408028?fullLog=true) failed without clear indication why.

Thanks for your work @HyukjinKwon @dongjoon-hyun and sorry for any misunderstandings on my side.

**Review by HyukjinKwon — APPROVED** [review]
Alright, then I am going to merge this. LGTM.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by HyukjinKwon** [comment]
@shaneknapp TR;DL: Now SparkR supports both testthat 1.0.2 and 2.0.0. I will backport this to branch-2.4 at #27379 as well.

So, as a reminder:
- We should upgrade testthat in Jenkins to 2.0.0 at any earliest time you're available
- We should upgrade R version to 3.4.x after Spark 3.0 release.

**Comment by shaneknapp** [comment]
@HyukjinKwon will do.  the upgrade won't happen until next week.

**Comment by shaneknapp** [comment]
> * We should upgrade testthat in Jenkins to 2.0.0 at any earliest time you're available

should be easy.

> * We should upgrade R version to 3.4.x after Spark 3.0 release.

won't be easy.  will be disruptive.
